src/main/java/nwoolcan/model/brewery/production/batch/BatchImpl.java

```diff
@@ -28,39 +28,73 @@
 
     private static final String CANNOT_CREATE_STEP_EXCEPTION = "Cannot create a step with the given type: ";
     private static final String CANNOT_FINALIZE_CURRENT_STEP = "Cannot finalize current step.";
+    private static final String BATCH_IS_ENDED_MESSAGE = "Cannot perform operation because batch is in ended state.";
+    private static final Object CANNOT_GO_TO_STEP_MESSAGE = "From this step, cannot go to step: ";
+    private static final String BATCH_NOT_ENDED_MESSAGE = "Cannot perform operation because batch is not in ended state.";
 
+    private final int id;
     private final ModifiableBatchInfo batchInfo;
     private final List<Step> steps;
 
     @Nullable
     private BatchEvaluation batchEvaluation;
 
     /**
-     * Batch Impl.
-     * @param beerDescription the description of the beer.
-     * @param batchMethod the method of the batch.
-     * @param initialSize the size of the batch.
-     * @param ingredients the ingredients for this batch.
-     * @param initialStep the starting point.
+     * Creates a new {@link Batch} in production.
+     * @param beerDescription the batch's beer description.
+     * @param batchMethod the batch's method.
+     * @param initialSize the initial size of the batch.
+     * @param ingredients the ingredients of the beer made by the batch.
+     * @param initialStep the initial step of the batch.
+     * @param waterMeasurement the water measurement of the batch.
+     * @throws IllegalArgumentException if the initial step cannot be created.
      */
     public BatchImpl(final BeerDescription beerDescription,
                      final BatchMethod batchMethod,
                      final Quantity initialSize,
                      final Collection<Pair<IngredientArticle, Quantity>> ingredients,
-                     final StepType initialStep) {
-        //TODO insert parameters
-        this.batchInfo = new ModifiableBatchInfoImpl(ingredients, beerDescription, batchMethod, initialSize);
+                     final StepType initialStep,
+                     @Nullable final WaterMeasurement waterMeasurement) {
+        this.id = BatchIdGenerator.getInstance().getNextId();
+
+        if (waterMeasurement == null) {
+            this.batchInfo = new ModifiableBatchInfoImpl(ingredients, beerDescription, batchMethod, initialSize);
+        } else {
+            this.batchInfo = new ModifiableBatchInfoImpl(ingredients, beerDescription, batchMethod, initialSize, waterMeasurement);
+        }
 
         final Result<Step> res = Steps.create(initialStep);
         if (res.isError()) {
             throw new IllegalArgumentException(CANNOT_CREATE_STEP_EXCEPTION + initialStep);
         }
 
-        res.getValue().addObserver(batchInfo);
+        res.getValue().addParameterObserver(batchInfo);
```

Comment:
better use `peek` probably

Reply:
same

Reply:
better use proper methods and there's no need to extract value from `Result`